src/theory/strings/array_solver.cpp

```diff
@@ -0,0 +1,241 @@
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Sequences solver for seq.nth/seq.update.
+ */
+
+#include "theory/strings/array_solver.h"
+
+#include "theory/strings/arith_entail.h"
+#include "theory/strings/theory_strings_utils.h"
+#include "util/rational.h"
+
+using namespace cvc5::context;
+using namespace cvc5::kind;
+
+namespace cvc5 {
+namespace theory {
+namespace strings {
+
+ArraySolver::ArraySolver(Env& env,
+                         SolverState& s,
+                         InferenceManager& im,
+                         TermRegistry& tr,
+                         CoreSolver& cs,
+                         ExtfSolver& es,
+                         ExtTheory& extt)
+    : EnvObj(env),
+      d_state(s),
+      d_im(im),
+      d_termReg(tr),
+      d_csolver(cs),
+      d_esolver(es),
+      d_eqProc(context())
+{
+  NodeManager* nm = NodeManager::currentNM();
+  d_zero = nm->mkConst(Rational(0));
+}
+
+ArraySolver::~ArraySolver() {}
+
+bool ArraySolver::isHandledUpdate(Node n)
+{
+  Assert(n.getKind() == STRING_UPDATE || n.getKind() == STRING_SUBSTR);
+  NodeManager* nm = NodeManager::currentNM();
+  Node lenN = n[2];
+  if (n.getKind() == STRING_UPDATE)
+  {
+    lenN = nm->mkNode(STRING_LENGTH, n[2]);
+  }
+  Node one = nm->mkConst(Rational(1));
+  return ArithEntail::checkEq(lenN, one);
+}
+
+Node ArraySolver::getUpdateBase(Node n)
+{
+  while (n.getKind() == STRING_UPDATE)
+  {
+    n = n[0];
+  }
+  return n;
+}
+
+void ArraySolver::checkArrayConcat()
+{
+  if (!d_termReg.hasSeqUpdate())
+  {
+    Trace("seq-array") << "No seq.update/seq.nth terms, skipping check..."
+                       << std::endl;
+    return;
+  }
+  d_currTerms.clear();
+  Trace("seq-array") << "ArraySolver::checkArrayConcat..." << std::endl;
+  checkTerms(STRING_UPDATE);
+  checkTerms(SEQ_NTH);
+}
+
+void ArraySolver::checkTerms(Kind k)
+{
+  Assert(k == STRING_UPDATE || k == SEQ_NTH);
+  NodeManager* nm = NodeManager::currentNM();
+  // get all the active update terms that have not been reduced in the
+  // current context by context-dependent simplification
+  std::vector<Node> terms = d_esolver.getActive(k);
+  for (const Node& t : terms)
+  {
+    Trace("seq-array-debug") << "check term " << t << "..." << std::endl;
+    Assert(t.getKind() == k);
+    if (k == STRING_UPDATE && !isHandledUpdate(t))
+    {
+      // not handled by procedure
+      Trace("seq-array-debug") << "...unhandled" << std::endl;
+      continue;
+    }
+    Node r = d_state.getRepresentative(t[0]);
+    NormalForm& nf = d_csolver.getNormalForm(r);
+    Trace("seq-array-debug") << "...normal form " << nf.d_nf << std::endl;
+    if (nf.d_nf.empty())
+    {
+      // should have been reduced (UPD_EMPTYSTR)
+      if (k == STRING_UPDATE) Assert(false);
```

Comment:
Nit: We always add braces for the blocks.
```suggestion
      if (k == STRING_UPDATE) { Assert(false); }
```